zerver/lib/mention.py

```diff
@@ -162,15 +171,52 @@ def get_stream_name_map(self, stream_names: Set[str]) -> Dict[str, int]:
                 .values(
                     "id",
                     "name",
+                    "recipient_id",
                 )
             )
 
             for row in rows:
                 self.stream_cache[row["name"]] = row["id"]
+                self.recipient_cache[row["name"]] = row["recipient_id"]
                 result[row["name"]] = row["id"]
 
         return result
 
+    def get_topic_info_map(
+        self, channel_topic: Set[ChannelTopicInfo]
+    ) -> Dict[ChannelTopicInfo, int]:
+        if not channel_topic:
+            return {}
+
+        result: Dict[ChannelTopicInfo, int] = {}
+        unseen_channel_topic: List[ChannelTopicInfo] = []
+
+        for channel_topic_object in channel_topic:
+            if channel_topic_object in self.topic_cache:
+                result[channel_topic_object] = self.topic_cache[channel_topic_object]
+            else:
+                unseen_channel_topic.append(channel_topic_object)
+
+        for channel_topic_object in unseen_channel_topic:
+            recipient_id = self.recipient_cache.get(channel_topic_object.channel_name)
+
+            assert recipient_id is not None
+
+            topic_latest_message = (
+                messages_for_topic(self.realm_id, recipient_id, channel_topic_object.topic_name)
+                .values_list("id", flat=True)
+                .order_by("-id")
+                .first()
+            )
```

Comment:
I think this construction could just be `.values_list(...).last()`:

https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet.first

Reply:
updated!